### PR TITLE
Support key composing (i.e. dead keys) in Wayland driver

### PR DESCRIPTION
**Pull request by lacc97**
Based on an old patch (#2488) from the old Bugzilla issue tracker. Unfortunately, migration to Github erased the name so I can't properly attribute authorship.

### CODE REVIEW

File: src/video/wayland/SDL_waylandevents.c [hunk 1]
```diff
@@ -718,15 +744,33 @@ keyboard_input_get_text(char text[8], const struct SDL_WaylandInput *input, uint
...
 #ifdef SDL_USE_IME
-    if (SDL_IME_ProcessKeyEvent(syms[0], key + 8)) {
+    if (SDL_IME_ProcessKeyEvent(sym, key + 8)) {
         *handled_by_ime = SDL_TRUE;
         return SDL_TRUE;
     }
 #endif
 
-    return WAYLAND_xkb_keysym_to_utf8(syms[0], text, 8) > 0;
+    if (WAYLAND_xkb_compose_state_feed(input->xkb.compose_state, syms[0]) == XKB_COMPOSE_FEED_ACCEPTED) {
```
Comment (flibitijibibo) [inline src/video/wayland/SDL_waylandevents.c:756]:
```suggestion
    if (WAYLAND_xkb_compose_state_feed(input->xkb.compose_state, sym) == XKB_COMPOSE_FEED_ACCEPTED) {
```

File: src/video/wayland/SDL_waylandevents.c [hunk 2]
```diff
@@ -661,6 +662,30 @@ keyboard_handle_keymap(void *data, struct wl_keyboard *keyboard,
...
+    /*
+     * See https://blogs.s-osg.org/compose-key-support-weston/
+     * for further explanation.
+     */
+
+    /* Look up the preferred locale, falling back to "C" as default */
+    if (!(locale = getenv("LC_ALL")))
+        if (!(locale = getenv("LC_CTYPE")))
+            if (!(locale = getenv("LANG")))
+                locale = "C";
```
Comment (flibitijibibo) [inline src/video/wayland/SDL_waylandevents.c:675]: We should probably replace this with SDL_GetLocale()

File: src/video/wayland/SDL_waylandevents.c [hunk 3]
```diff
@@ -661,6 +662,30 @@ keyboard_handle_keymap(void *data, struct wl_keyboard *keyboard,
         input->xkb.keymap = NULL;
         return;
     }
+
+    /*
+     * See https://blogs.s-osg.org/compose-key-support-weston/
+     * for further explanation.
```
Comment (flibitijibibo) [inline src/video/wayland/SDL_waylandevents.c:668]:
```suggestion
     * for further explanation on dead keys in Wayland.
```

**Review by flibitijibibo — CHANGES_REQUESTED** [review]
Excellent work on rebasing this! I dug into the web archive and unfortunately the only name I could dig up was "chw", no e-mail address or anything. It does look like this was taken from that Weston URL though, so I dunno...

In any case, this is just what I would have reviewed for the original patch.

**Comment by sezero** [comment]
Original bugzilla entry is https://bugzilla.libsdl.org/show_bug.cgi?id=3687
https://web.archive.org/web/20210210031224/https://bugzilla.libsdl.org/show_bug.cgi?id=3687
says the user email name is 'chw' - I can't get anything more out of it.

File: src/video/wayland/SDL_waylandevents.c (hunk 2 above)
Comment (lacc97) [reply src/video/wayland/SDL_waylandevents.c:675]: I'm not sure how to convert between `SDL_Locale` to the locale string format expected by the xkb functions. I checked the x11 driver code and it seems to just directly use the relevant environment variables.

Comment (flibitijibibo) [reply src/video/wayland/SDL_waylandevents.c:675]: Ah, fair enough then - in that case, let's just replace `getenv` with `SDL_getenv` for stdlib compatibility and leave it at that.

**Review by flibitijibibo — APPROVED** [review]
Latest lgtm! That said I am absolutely clueless on how dead keys actually work, so as long as someone who knows what they're doing can verify that this is functional this is good to go.

**Comment by lacc97** [comment]
Just for reference, here's the xkbcommon documentation on compose keys: https://xkbcommon.org/doc/current/group__compose.html

**Comment by flibitijibibo** [comment]
I tested this for a bit with my admittedly little knowledge of compose keys and was able to verify that this works as advertised. Should be good to merge.

File: src/video/wayland/SDL_waylandevents.c [hunk 4]
```diff
@@ -661,6 +662,30 @@ keyboard_handle_keymap(void *data, struct wl_keyboard *keyboard,
         input->xkb.keymap = NULL;
         return;
     }
+
+    /*
+     * See https://blogs.s-osg.org/compose-key-support-weston/
+     * for further explanation on dead keys in Wayland.
+     */
```
Comment (flibitijibibo) [inline src/video/wayland/SDL_waylandevents.c:669]: Looks like this link is pretty dead even on Wayback Machine, so we can probably replace this with the official documentation for dead keys.

File: src/video/wayland/SDL_waylandevents.c [hunk 5]
```diff
@@ -661,6 +662,30 @@ keyboard_handle_keymap(void *data, struct wl_keyboard *keyboard,
...
+
+    /* Look up the preferred locale, falling back to "C" as default */
+    if (!(locale = SDL_getenv("LC_ALL")))
+        if (!(locale = SDL_getenv("LC_CTYPE")))
+            if (!(locale = SDL_getenv("LANG")))
+                locale = "C";
```
Comment (flibitijibibo) [inline src/video/wayland/SDL_waylandevents.c:675]:
```suggestion
    if (!(locale = SDL_getenv("LC_ALL"))) {
        if (!(locale = SDL_getenv("LC_CTYPE"))) {
            if (!(locale = SDL_getenv("LANG"))) {
                locale = "C";
            }
        }
    }
```

**Review by flibitijibibo — COMMENTED** [review]
This got stomped on a bit by #4303, so it needs a quick rebase. Since we're having to rebase, figured I'd do one more pass for any possible nits.

**Pull request merged**